<a href="https://colab.research.google.com/github/GeoLabUniLaSalle/Python/blob/main/Les_fichiers_externes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Les fichiers externes**

Dans ce chapitre, nous allons voir comment ouvrir un fichier externe pour accéder ou modifier son contenu.

Débutons par la création d'un fichier dans lequel nous allons écrire du texte.

In [ ]:
fichier = open("data.txt", "a") # a indique que le fichier est ouvert en écriture, ajoutant du contenu à la fin du fichier s'il existe
fichier.write("Hello World !")
fichier.close()

Le fichier est accessible dans les fichiers de ce Jupyter Notebook.

Voici les différents modes d'ouverture disponibles.

In [ ]:
help(open)

Help on built-in function open in module io:

open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)
    Open file and return a stream.  Raise OSError upon failure.
    
    file is either a text or byte string giving the name (and the path
    if the file isn't in the current working directory) of the file to
    be opened or an integer file descriptor of the file to be
    wrapped. (If a file descriptor is given, it is closed when the
    returned I/O object is closed, unless closefd is set to False.)
    
    mode is an optional string that specifies the mode in which the file
    is opened. It defaults to 'r' which means open for reading in text
    mode.  Other common values are 'w' for writing (truncating the file if
    it already exists), 'x' for creating and writing to a new file, and
    'a' for appending (which on some Unix systems, means that all writes
    append to the end of the file regardless of the current seek position

Ecrivons maintenant le contenu d'une liste de listes (équivalent à un tableau à 2 dimensions) dans un fichier externe.

In [ ]:
l=[[1, 3], [2, 4]]
fichier = open("data.txt", "w") # w indique que le fichier est ouvert en écriture, écrasant le contenu du fichier
for x in l :
  for y in x :
    fichier.write(str(y))
fichier.close()

Dans le sens inverse, accédons au contenu du fichier.

In [ ]:
fichier = open("data.txt", "r") # r indique que le fichier est ouvert en lecture (mode par défaut)
z = fichier.read()
fichier.close()
print(z, type(z))

1324 <class 'str'>


Le fichier contient le texte "1324". Nous n'avons pas stocké la structure de la liste (liste de listes).

Stockons maintenant le contenu de notre liste dans un fichier au format csv.

In [ ]:
import csv
l=[[1, 3], [2, 4]]
with open('data.csv', 'w', newline='') as fichier:
    writer = csv.writer(fichier)
    writer.writerows(l)

Ajoutons des en-têtes de colonnes.

In [1]:
import csv
l=[[1, 3], [2, 4]]
champs = ['col1','col2']
with open('data.csv', 'w', newline='') as fichier:
    writer = csv.writer(fichier)
    writer.writerow(champs)
    writer.writerows(l)

Accédons maintenant au contenu de ce fichier.

In [2]:
import csv
with open('data.csv', 'r') as fichier:
    reader = csv.reader(fichier)
    for i in reader:
        print(i)

['col1', 'col2']
['1', '3']
['2', '4']


Stockons ce contenu dans une nouvelle liste équivalente à la liste initiale.

In [7]:
import csv
with open('data.csv', newline='') as fichier:
    reader = csv.reader(fichier)
    headings = next(reader)
    l2 = list(reader)
print(l2)

[['1', '3'], ['2', '4']]


Pour convertir les str en int, nous pouvons utiliser la fonction map qui applique un opérateur sur tous les termes d'une liste. Ici, nous appliquons l'operateur de conversion 'int' sur tous les termes de la liste x.

In [ ]:
l3=[]
for x in l2:
  l3.append(list(map(int,x)))
print(l3)

[[1, 3], [2, 4]]


Une autre fonction souvent utile lorsque l'on traite des fichiers csv est la fonction strip qui élimine des caractères inutiles. Par défaut, strip supprime les caractères 'espace' qui sont en début et fin d'un str.

In [ ]:
x = '     bonjour     '
print(x.strip())

bonjour


Il est possible de spécifier certains caractères à éliminer, mais toujours en début ou fin de str.

In [ ]:
x = '!!???###bonjour /?!#\ bonjour##??!!!!'
print(x.strip('?#!'))

bonjour /?!#\ bonjour


Il est possible de créer un fichier .zip contenant plusieurs fichiers.

In [13]:
from zipfile import ZipFile

zipObj = ZipFile('data.zip', 'w')

for i in range(3):
  nom = "a"+str(i)+".txt"
  fichier = open(nom,"w")
  fichier.write("Hello")
  fichier.close()
  zipObj.write(nom)

zipObj.close()